client/allocrunner/alloc_runner.go

```diff
@@ -176,40 +176,53 @@ func (ar *allocRunner) WaitCh() <-chan struct{} {
 	return ar.waitCh
 }
 
-// Run is the main goroutine that executes all the tasks.
+// Run the AllocRunner. Starts tasks if the alloc is non-terminal and closes
+// WaitCh when it exits. Should be started in a goroutine.
 func (ar *allocRunner) Run() {
-	ar.destroyedLock.Lock()
-	defer ar.destroyedLock.Unlock()
+	// Close the wait channel on return
+	defer close(ar.waitCh)
 
-	// Run should not be called after Destroy is called. This is a
-	// programming error.
-	if ar.destroyed {
-		ar.logger.Error("alloc destroyed; cannot run")
-		return
-	}
+	// Start the task state update handler
+	go ar.handleTaskStateUpdates()
 
 	// If an alloc should not be run, ensure any restored task handles are
 	// destroyed and exit to wait for the AR to be GC'd by the client.
 	if !ar.shouldRun() {
 		ar.logger.Debug("not running terminal alloc")
 
-		// Cleanup and sync state
-		states := ar.killTasks()
-
-		// Get the client allocation
-		calloc := ar.clientAlloc(states)
+		// Ensure all tasks are cleaned up
+		ar.killTasks()
+		return
+	}
 
-		// Update the server
-		ar.stateUpdater.AllocStateUpdated(calloc)
+	// Mark task runners as being run for Shutdown
+	ar.destroyedLock.Lock()
+	ar.runnersLaunched = true
+	ar.destroyedLock.Unlock()
 
-		// Broadcast client alloc to listeners
-		ar.allocBroadcaster.Send(calloc)
+	// If task update chan has been closed, that means we've been shutdown.
+	select {
+	case <-ar.taskStateUpdateHandlerCh:
 		return
+	default:
+	}
+
+	// Run the prestart hooks
+	if err := ar.prerun(); err != nil {
+		ar.logger.Error("prerun failed", "error", err)
+		goto POST
+	}
+
+	// Run the runners (blocks until they exit)
+	ar.runTasks()
+
+POST:
+	// Run the postrun hooks
+	// XXX Equivalent to TR.Poststop hook
```

Comment:
What does this comment mean?

Reply:
We goofed up the names: AR post *run* hooks ~= TR post *stop* hooks.

I keep meaning to rename them for consistency, but I don't want to muddle these already muddled PRs.